src/tqec/computation/pipe.py

```diff
@@ -0,0 +1,194 @@
+        return Direction3D(str(self).index("O"))
+
+    def get_basis_along(
+        self, direction: Direction3D, at_head: bool = True
+    ) -> ZXBasis | None:
+        """Get the basis of the kind in the given direction.
+
+        Args:
+            direction: The direction of the basis.
+            at_head: If True, get the basis at the head side of the pipe. Otherwise,
+                get the basis at the tail of the pipe. The head side will have the
+                opposite basis as the tail if the pipe has hadamard transition.
+
+        Returns:
+            None if the basis is not defined in the direction, i.e. the direction is
+            the same as the pipe direction. Otherwise, the basis in the direction at
+            the specified side.
+        """
+        if direction == self.direction:
+            return None
+        head_basis = ZXBasis(str(self)[direction.value])
+        if not at_head and self.has_hadamard:
+            return head_basis.with_zx_flipped()
+        return head_basis
+
+    @property
+    def is_temporal(self) -> bool:
+        """Check if the pipe is temporal."""
+        return self.z is None
+
+    @property
+    def is_spatial(self) -> bool:
+        """Check if the pipe is spatial."""
+        return not self.is_temporal
+
+    @staticmethod
+    def from_cube_kind(
+        cube_kind: ZXCube,
+        direction: Direction3D,
+        cube_at_head: bool,
+        has_hadamard: bool = False,
+    ) -> PipeKind:
+        """Infer the pipe kind from the endpoint cube kind and the pipe direction."""
+        bases: list[ZXBasis]
+        if not cube_at_head and has_hadamard:
+            bases = [basis.with_zx_flipped() for basis in cube_kind.as_tuple()]
+        else:
+            bases = list(cube_kind.as_tuple())
+        bases_str = [basis.value for basis in bases]
+        bases_str[direction.value] = "O"
+        if has_hadamard:
+            bases_str.append("H")
+        return PipeKind.from_str("".join(bases_str))
+
+
+@dataclass(frozen=True)
+class Pipe:
+    """A block connecting two cubes in a 3D spacetime diagram.
+
+    Attributes:
+        u: The first cube. The position of u will be guaranteed to be less than v.
+        v: The second cube. The position of v will be guaranteed to be greater than u.
+        kind: The type of the pipe.
+    """
+
+    u: Cube
+    v: Cube
+    kind: PipeKind
+
+    def __post_init__(self) -> None:
+        u, v = self.u, self.v
+        p1, p2 = u.position, v.position
+        shift = [0, 0, 0]
+        shift[self.kind.direction.value] = 1
+        if not p1.shift_by(*shift) == p2 and not p2.shift_by(*shift) == p1:
+            raise TQECException(
+                f"The pipe must connect two nearby cubes in direction {self.kind.direction}."
+            )
+
+        # Ensure position of u is less than v
+        if p1 > p2:
+            object.__setattr__(self, "u", v)
+            object.__setattr__(self, "v", u)
```

Comment:
I do not know what is better between this approach and [raising an error on non-sorted cubes and providing a static method to build a `Pipe` from `Cube`s that would sort the cubes correctly]. Changing the cubes order might be confusing for the user, but it might also not be important in the end depending how the class is used.